File: LeetSync/1522-stone-game-iii/stone-game-iii.py
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)

        dp = [0] * 4

        for i in range(n - 1, -1, -1):
            max_diff = float("-inf")
            current_take = 0

            for k in range(1, 4):
                if i + k <= n:
                    current_take += stoneValue[i + k - 1]
                    max_diff = max(max_diff, current_take - dp[(i + k) % 4])

            dp[i % 4] = max_diff

        score_diff = dp[0]
        if score_diff > 0:
            return "Alice"
        elif score_diff < 0:
            return "Bob"
        else:
            return "Tie"

File: LeetSync/1522-stone-game-iii/stone-game-iii.cpp
class Solution {
public:
    string stoneGameIII(vector<int>& stoneValue) {
        int n = stoneValue.size();
        
        vector<int> dp(4, 0);

        for (int i = n - 1; i >= 0; --i) {
            int maxDiff = INT_MIN;
            int currentTake = 0;

            for (int k = 1; k <= 3 && i + k <= n; ++k) {
                currentTake += stoneValue[i + k - 1];
                maxDiff = std::max(maxDiff, currentTake - dp[(i + k) % 4]);
            }
            
            dp[i % 4] = maxDiff;
        }

        int scoreDiff = dp[0];
        if (scoreDiff > 0) return "Alice";
        if (scoreDiff < 0) return "Bob";
        return "Tie";
    }
};